lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource'
+require 'kubernetes-deploy/kubeclient_builder'
+require "active_record"
+require "jsonpath"
+
+module KubernetesDeploy
+  class DiscoverableResource < KubernetesResource
+    extend KubernetesDeploy::KubeclientBuilder
+
+    class << self
+      attr_accessor :version, :group, :type
+      attr_reader :timeout
+    end
+
+    STATUS_FIELD_ANNOTATION = 'kubernetes-deploy.shopify.io/status-field'
+    STATUS_SUCCESS_ANNOTATION = 'kubernetes-deploy.shopify.io/status-success'
+    TIMEOUT_ANNOTATION = 'kubernetes-deploy.shopify.io/timeout'
+    PREDEPLOY_ANNOTATION = 'kubernetes-deploy.shopify.io/predeploy'
+    PRUNABLE_ANNOTATION = 'kubernetes-deploy.shopify.io/prunable'
+
+    def type
+      self.class.type
+    end
+
+    def self.inherited(child_class)
+      child_classes.add child_class
+    end
+
+    def self.child_classes
+      @child_classes ||= Set.new
+    end
+
+    def self.prunable?
```

Comment:
Could we add this and `identity` to the other resources classes and build the _entire_ prune list dynamically? It would be awesome if we could be dynamically generating 99% "normal" subclasses of `KuberenetesResource` rather than subtly different classes.